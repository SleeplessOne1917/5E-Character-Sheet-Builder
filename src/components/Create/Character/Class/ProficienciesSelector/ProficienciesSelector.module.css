.container {
    text-align: center;
}

.proficiencies-container {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
    align-items: center;
}

.types-container {
    display: flex;;
    align-items: center;
    justify-content: center;
}

.proficiency-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    padding: 1rem;
    background-color: var(--heaven-light);
    border: 3px solid var(--davy);
    border-radius: 1rem;
    width: fit-content;
    margin: 1rem;
}

.selected {
    background-color: var(--highlight) !important;
}

.selected-text {
    font-size: 1.5rem;
    text-align: inherit;
}

.label {
    font-size: 1.3rem;
    padding: 0.5rem;
}

.heading {
    align-self: center;
    margin-block-start: 3rem;
    margin-block-end: 3rem;
}