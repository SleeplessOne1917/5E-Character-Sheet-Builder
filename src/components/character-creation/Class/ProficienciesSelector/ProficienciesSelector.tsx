import { useCallback, useEffect, useMemo, useState } from 'react';
import { useAppDispatch, useAppSelector } from '../../../../hooks/reduxHooks';
import { setClassProficiencies } from '../../../../redux/features/classInfo';
import {
	addProficiency,
	removeProficiency
} from '../../../../redux/features/proficiencies';
import { getProficiencyTypeName } from '../../../../services/proficiencyTypeService';
import {
	ProficiencyType,
	SrdProficiencyItem,
	SrdProficiencyItemChoice
} from '../../../../types/srd';
import Select, { Option } from '../../../Select/Select';
import classes from './ProficienciesSelector.module.css';

type ProficienciesSelectorProps = {
	choice: SrdProficiencyItemChoice;
	proficienciesIndex: number;
};

const ProficienciesSelector = ({
	choice,
	proficienciesIndex
}: ProficienciesSelectorProps) => {
	const classProficienciesList = useAppSelector(
		state => state.editingCharacter.classInfo.proficiencies
	);
	const allProficiencies = useAppSelector(
		state => state.editingCharacter.proficiencies
	);

	const dispatch = useAppDispatch();

	const [selectedType, setSelectedType] = useState('blank');

	const types = useMemo(() => {
		const types: Option[] = [];
		const hasNestedChoices = choice.from.options.some(option => option.choice);

		if (hasNestedChoices) {
			for (const option of choice.from.options) {
				const type = option.choice?.from.options[0].item
					?.type as ProficiencyType;

				types.push({ value: type, label: getProficiencyTypeName(type) });
			}
		}

		return types;
	}, [choice.from.options]);

	const values = useMemo(() => {
		if (
			classProficienciesList.length > proficienciesIndex &&
			classProficienciesList[proficienciesIndex].length > 0
		) {
			return classProficienciesList[proficienciesIndex].map(
				prof => prof?.index ?? null
			);
		}

		let choose: number;
		if (selectedType !== 'blank') {
			choose = choice.from.options.find(option =>
				option.choice?.from.options.some(o => o.item?.type === selectedType)
			)?.choice?.choose as number;
		} else {
			choose = choice.choose;
		}

		const values: (string | null)[] = [];

		for (let i = 0; i < choose; ++i) {
			values.push(null);
		}

		return values;
	}, [
		choice.choose,
		classProficienciesList,
		choice.from.options,
		proficienciesIndex,
		selectedType
	]);

	useEffect(() => {
		if (!classProficienciesList[proficienciesIndex]) {
			dispatch(
				setClassProficiencies({
					index: proficienciesIndex,
					proficiencies: values as null[]
				})
			);
		}
	}, [dispatch, proficienciesIndex, values, classProficienciesList]);

	const typeString = useMemo(() => {
		if (selectedType !== 'blank') {
			return types.find(({ value }) => value === selectedType)?.label as string;
		} else if (types.length > 0) {
			return 'Proficiencies';
		} else {
			return getProficiencyTypeName(
				choice.from.options[0].item?.type as ProficiencyType
			);
		}
	}, [choice.from.options, selectedType, types]);

	const proficiencies = useMemo(() => {
		let proficiencies: SrdProficiencyItem[] = [];

		if (selectedType !== 'blank') {
			proficiencies = choice.from.options
				.find(option =>
					option.choice?.from.options.some(o => o.item?.type === selectedType)
				)
				?.choice?.from.options.map(({ item }) => item) as SrdProficiencyItem[];
		} else if (types.length === 0) {
			proficiencies = choice.from.options.map(
				({ item }) => item as SrdProficiencyItem
			);
		}

		return proficiencies;
	}, [choice.from.options, selectedType, types]);

	const handleChange = useCallback(
		(index: number, value: SrdProficiencyItem | null) => {
			if (value) {
				dispatch(addProficiency(value));
			} else if (values[index]) {
				dispatch(removeProficiency(values[index] as string));
			}

			dispatch(
				setClassProficiencies({
					index: proficienciesIndex,
					proficiencies: [
						...classProficienciesList[proficienciesIndex].slice(0, index),
						value,
						...classProficienciesList[proficienciesIndex].slice(index + 1)
					]
				})
			);
		},
		[dispatch, classProficienciesList, proficienciesIndex, values]
	);

	const selectors = useMemo(
		() =>
			values.map((value, index) => (
				<div
					className={`${classes['proficiency-container']}${
						value ? ` ${classes.selected}` : ''
					}`}
					key={index}
				>
					<label id="select-proficiency" className={classes.label}>
						Select{' '}
						{typeString.endsWith('s') ? typeString.slice(0, -1) : typeString}
					</label>
					<Select
						labelledBy="select-proficiency"
						value={value ? value : 'blank'}
						options={[{ value: 'blank', label: '\u2014' } as Option].concat(
							proficiencies
								.filter(
									prof =>
										prof.index === value ||
										!(
											allProficiencies.some(p => p.index === prof.index) ||
											values.includes(prof.index)
										)
								)
								.map(prof => ({
									value: prof.index,
									label: prof.name.replace(/.*:\s+/, '')
								}))
						)}
						onChange={value =>
							handleChange(
								index,
								value !== 'blank'
									? (proficiencies.find(
											prof => prof.index === (value as string)
									  ) as SrdProficiencyItem)
									: null
							)
						}
					/>
				</div>
			)),
		[allProficiencies, handleChange, proficiencies, typeString, values]
	);

	const handleSelectType = useCallback(
		(type: ProficiencyType | 'blank') => {
			setSelectedType(type);
			dispatch(
				setClassProficiencies({
					index: proficienciesIndex,
					proficiencies: []
				})
			);
		},
		[setSelectedType, dispatch, proficienciesIndex]
	);

	return (
		<div data-testid="proficiencies-selector" className={classes.container}>
			<h3 className={classes.heading}>{typeString}</h3>
			<div className={classes['selected-text']}>
				{values.filter(skill => skill).length}/{values.length} {typeString}{' '}
				Selected
			</div>
			{types.length > 0 ? (
				<>
					<div className={classes['types-container']}>
						<div
							className={`${classes['proficiency-container']}${
								selectedType !== 'blank' ? ` ${classes.selected}` : ''
							}`}
						>
							<label id="select-proficiency-type" className={classes.label}>
								Select Proficiency Type
							</label>
							<Select
								options={[{ value: 'blank', label: '\u2014' } as Option].concat(
									types
								)}
								value={selectedType}
								onChange={value =>
									handleSelectType(value as ProficiencyType | 'blank')
								}
							/>
						</div>
					</div>
					{selectedType !== 'blank' && (
						<div className={classes['proficiencies-container']}>
							{selectors}
						</div>
					)}
				</>
			) : (
				<div className={classes['proficiencies-container']}>{selectors}</div>
			)}
		</div>
	);
};

export default ProficienciesSelector;
