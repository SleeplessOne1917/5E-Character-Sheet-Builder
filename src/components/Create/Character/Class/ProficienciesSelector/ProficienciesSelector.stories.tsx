import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Provider } from 'react-redux';
import { getTestStore } from '../../../../../redux/store';
import ProficienciesSelector from './ProficienciesSelector';

export default {
	title: 'Components/CharacterCreation/Class/ProficienciesSelector',
	component: ProficienciesSelector,
	args: {
		proficienciesIndex: 0
	},
	decorators: [
		Story => (
			<Provider store={getTestStore()}>
				<Story />
			</Provider>
		)
	]
} as ComponentMeta<typeof ProficienciesSelector>;

const Template: ComponentStory<typeof ProficienciesSelector> = args => (
	<ProficienciesSelector {...args} />
);

export const Default = Template.bind({});
Default.args = {
	choice: {
		choose: 2,
		from: {
			options: [
				{
					item: {
						index: 'skills-acrobatics',
						name: 'Skill: Acrobatics',
						type: 'SKILLS'
					}
				},
				{
					item: {
						index: 'skills-athletics',
						name: 'Skill: Athletics',
						type: 'SKILLS'
					}
				},
				{
					item: {
						index: 'skills-history',
						name: 'Skill: History',
						type: 'SKILLS'
					}
				},
				{
					item: {
						index: 'skills-insight',
						name: 'Skill: Insight',
						type: 'SKILLS'
					}
				},
				{
					item: {
						index: 'skills-religion',
						name: 'Skill: Religion',
						type: 'SKILLS'
					}
				},
				{
					item: {
						index: 'skills-stealth',
						name: 'Skill: Stealth',
						type: 'SKILLS'
					}
				}
			]
		}
	}
};

export const WithTypes = Template.bind({});
WithTypes.args = {
	choice: {
		choose: 1,
		from: {
			options: [
				{
					choice: {
						choose: 1,
						from: {
							options: [
								{
									item: {
										index: 'alchemists-supplies',
										name: "Alchemist's Supplies",
										type: 'ARTISANS_TOOLS'
									}
								},
								{
									item: {
										index: 'brewers-supplies',
										name: "Brewer's Supplies",
										type: 'ARTISANS_TOOLS'
									}
								},
								{
									item: {
										index: 'calligraphers-supplies',
										name: "Calligrapher's Supplies",
										type: 'ARTISANS_TOOLS'
									}
								},
								{
									item: {
										index: 'carpenters-tools',
										name: "Carpenter's Tools",
										type: 'ARTISANS_TOOLS'
									}
								},
								{
									item: {
										index: 'cartographers-tools',
										name: "Cartographer's Tools",
										type: 'ARTISANS_TOOLS'
									}
								},
								{
									item: {
										index: 'cobblers-tools',
										name: "Cobbler's Tools",
										type: 'ARTISANS_TOOLS'
									}
								}
							]
						}
					}
				},
				{
					choice: {
						choose: 1,
						from: {
							options: [
								{
									item: {
										index: 'bagpipes',
										name: 'Bagpipes',
										type: 'MUSICAL_INSTRUMENTS'
									}
								},
								{
									item: {
										index: 'drum',
										name: 'Drum',
										type: 'MUSICAL_INSTRUMENTS'
									}
								},
								{
									item: {
										index: 'dulcimer',
										name: 'Dulcimer',
										type: 'MUSICAL_INSTRUMENTS'
									}
								},
								{
									item: {
										index: 'flute',
										name: 'Flute',
										type: 'MUSICAL_INSTRUMENTS'
									}
								},
								{
									item: {
										index: 'lute',
										name: 'Lute',
										type: 'MUSICAL_INSTRUMENTS'
									}
								},
								{
									item: {
										index: 'lyre',
										name: 'Lyre',
										type: 'MUSICAL_INSTRUMENTS'
									}
								}
							]
						}
					}
				}
			]
		}
	}
};
