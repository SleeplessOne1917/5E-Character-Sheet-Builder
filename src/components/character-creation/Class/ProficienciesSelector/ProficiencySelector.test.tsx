import { render, screen } from '@testing-library/react';
import * as stories from './ProficienciesSelector.stories';
import { composeStories } from '@storybook/testing-react';

const { Default, WithTypes } = composeStories(stories);

describe('renders correctly', () => {
	it('when default', () => {
		render(<Default />);

		expect(screen.getByTestId('proficiencies-selector')).toMatchSnapshot();
	});

	it('with types', () => {
		render(<WithTypes />);

		expect(screen.getByTestId('proficiencies-selector')).toMatchSnapshot();
	});
});
