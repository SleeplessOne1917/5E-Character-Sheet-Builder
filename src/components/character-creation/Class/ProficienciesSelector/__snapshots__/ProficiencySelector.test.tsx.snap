// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when default 1`] = `
<div
  class="container"
  data-testid="proficiencies-selector"
>
  <h3
    class="heading"
  >
    Skills
  </h3>
  <div
    class="selected-text"
  >
    0
    /
    2
     
    Skills
     
    Selected
  </div>
  <div
    class="proficiencies-container"
  >
    <div
      class="proficiency-container"
    >
      <div
        class="list-box-and-error-container"
        data-testid="select"
      >
        <label
          id="select-proficiency-label"
          style="font-size: 1.3rem;"
        >
          Select Skill
        </label>
        <div
          aria-labelledby="select-proficiency-label"
          class="list-box-container"
          style="font-size: 1.2rem;"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="button"
            data-testid="select-button"
            type="button"
          >
            —
            <svg
              aria-hidden="true"
              class="icon"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <ul
            aria-activedescendant="—"
            class="list"
            data-testid="select-list"
            role="listbox"
            style="display: none;"
            tabindex="-1"
          >
            <li
              aria-selected="true"
              class="item selected"
              data-testid="data-blank"
              role="option"
              tabindex="0"
            >
              —
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-acrobatics"
              role="option"
              tabindex="0"
            >
              Acrobatics
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-athletics"
              role="option"
              tabindex="0"
            >
              Athletics
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-history"
              role="option"
              tabindex="0"
            >
              History
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-insight"
              role="option"
              tabindex="0"
            >
              Insight
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-religion"
              role="option"
              tabindex="0"
            >
              Religion
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-stealth"
              role="option"
              tabindex="0"
            >
              Stealth
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="proficiency-container"
    >
      <div
        class="list-box-and-error-container"
        data-testid="select"
      >
        <label
          id="select-proficiency-label"
          style="font-size: 1.3rem;"
        >
          Select Skill
        </label>
        <div
          aria-labelledby="select-proficiency-label"
          class="list-box-container"
          style="font-size: 1.2rem;"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="button"
            data-testid="select-button"
            type="button"
          >
            —
            <svg
              aria-hidden="true"
              class="icon"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <ul
            aria-activedescendant="—"
            class="list"
            data-testid="select-list"
            role="listbox"
            style="display: none;"
            tabindex="-1"
          >
            <li
              aria-selected="true"
              class="item selected"
              data-testid="data-blank"
              role="option"
              tabindex="0"
            >
              —
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-acrobatics"
              role="option"
              tabindex="0"
            >
              Acrobatics
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-athletics"
              role="option"
              tabindex="0"
            >
              Athletics
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-history"
              role="option"
              tabindex="0"
            >
              History
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-insight"
              role="option"
              tabindex="0"
            >
              Insight
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-religion"
              role="option"
              tabindex="0"
            >
              Religion
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-skills-stealth"
              role="option"
              tabindex="0"
            >
              Stealth
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with types 1`] = `
<div
  class="container"
  data-testid="proficiencies-selector"
>
  <h3
    class="heading"
  >
    Proficiencies
  </h3>
  <div
    class="selected-text"
  >
    0
    /
    1
     
    Proficiencies
     
    Selected
  </div>
  <div
    class="types-container"
  >
    <div
      class="proficiency-container"
    >
      <label
        class="label"
        id="select-proficiency-type"
      >
        Select Proficiency Type
      </label>
      <div
        class="list-box-and-error-container"
        data-testid="select"
      >
        <div
          aria-labelledby="select-label"
          class="list-box-container"
          style="font-size: 1.2rem;"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="button"
            data-testid="select-button"
            type="button"
          >
            —
            <svg
              aria-hidden="true"
              class="icon"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <ul
            aria-activedescendant="—"
            class="list"
            data-testid="select-list"
            role="listbox"
            style="display: none;"
            tabindex="-1"
          >
            <li
              aria-selected="true"
              class="item selected"
              data-testid="data-blank"
              role="option"
              tabindex="0"
            >
              —
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-ARTISANS_TOOLS"
              role="option"
              tabindex="0"
            >
              Artisan's Tools
            </li>
            <li
              aria-selected="false"
              class="item"
              data-testid="data-MUSICAL_INSTRUMENTS"
              role="option"
              tabindex="0"
            >
              Musical Instruments
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
